walletnode/services/userdataprocess/task.go

```diff
@@ -0,0 +1,398 @@
+	"github.com/pastelnetwork/gonode/common/errors"
+	"github.com/pastelnetwork/gonode/common/log"
+	"github.com/pastelnetwork/gonode/common/service/task"
+	"github.com/pastelnetwork/gonode/common/service/userdata"
+	"github.com/pastelnetwork/gonode/walletnode/services/userdataprocess/node"
+)
+
+// Task is the task of userdata processing.
+type Task struct {
+	task.Task
+	*Service
+
+	// information of nodes process to set userdata
+	nodes      node.List
+	resultChan chan *userdata.UserdataProcessResult
+	err        error
+	request    *userdata.UserdataProcessRequest
+
+	// information user pastelid to retrieve userdata
+	userpastelid  string // user pastelid
+	resultChanGet chan *userdata.UserdataProcessRequest
+}
+
+// Run starts the task
+func (task *Task) Run(ctx context.Context) error {
+	ctx = log.ContextWithPrefix(ctx, fmt.Sprintf("%s-%s", logPrefix, task.ID()))
+
+	log.WithContext(ctx).Debugf("Start task")
+	defer log.WithContext(ctx).Debugf("End task")
+	defer close(task.resultChan)
+	defer close(task.resultChanGet)
+
+	if err := task.run(ctx); err != nil {
+		task.err = err
+		task.UpdateStatus(StatusTaskFailure)
+		log.WithContext(ctx).WithError(err).Warnf("Task failed")
+
+		return nil
+	}
+
+	task.UpdateStatus(StatusTaskCompleted)
+
+	return nil
+}
+
+func (task *Task) run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	maxNode := task.config.NumberSuperNodes
+	if task.request == nil {
+		// This is process to retrieve userdata
+		maxNode = 1 // Get data from 1 supernode only, currently we choose the 1st ranked supernode, but this may change later
+	}
+
+	// TODO: Make this init and connect to super nodes to generic reusable function to avoid code duplication (1)
+	// Retrieve supernodes with highest ranks.
+	topNodes, err := task.pastelTopNodes(ctx, maxNode)
+	if err != nil {
+		return err
+	}
+	if len(topNodes) < maxNode {
+		task.UpdateStatus(StatusErrorNotEnoughMasterNode)
+		return errors.New("unable to find enough Supernodes to send userdata to")
+	}
+
+	// Try to create mesh of supernodes, connecting to all supernodes in a different sequences.
+	var nodes node.List
+	var errs error
+	nodes, err = task.meshNodes(ctx, topNodes, 0) // Connect a mesh node with primary is 1st ranked SN
+	if err != nil {
+		if errors.IsContextCanceled(err) {
+			return err
+		}
+		errs = errors.Append(errs, err)
+		log.WithContext(ctx).WithError(err).Warnf("Could not create a mesh of the nodes")
+	}
+
+	if len(nodes) < maxNode {
+		return errors.Errorf("Could not create a mesh of %d nodes: %w", task.config.NumberSuperNodes, errs)
+	}
+
+	// Activate supernodes that are in the mesh.
+	nodes.Activate()
+	// Disconnect supernodes that are not involved in the process.
+	topNodes.DisconnectInactive()
+
+	// Cancel context when any connection is broken.
+	groupConnClose, _ := errgroup.WithContext(ctx)
+	groupConnClose.Go(func() error {
+		defer cancel()
+		return nodes.WaitConnClose(ctx)
```

Comment:
What is happen if `nodes.WaitConnClose(ctx)` return error ?

I think better, we should use a `chan` to `WaitConnClose`.
refer to [link](https://github.com/pastelnetwork/gonode/blob/623c2da2349ba98f58494cee2af3121e1167aefb/walletnode/services/artworkregister/task.go#L109)

Reply:
Right now this function is reused from multiple workflow so need to to refactor it for all the workflow and will do in another task.